test_cases/stdlib/check_dataclasses.py

```diff
@@ -88,3 +88,14 @@ def check_other_isdataclass_overloads(x: type, y: object) -> None:
         assert_type(dc.asdict(y), Dict[str, Any])
         assert_type(dc.astuple(y), Tuple[Any, ...])
         dc.replace(y)
+
+
+# Regression test for #11653
+D = dc.make_dataclass(
+    "D", [("a", int | None), "y", ("z", Annotated[frozenset[bytes], "metadata"], dc.field(default=frozenset()))]
+)
+# Check that it's inferred by the type checker as a class object of some kind
+# (but don't assert the exact type that `D` is inferred as,
+# in case a type checker decides to add some special-casing for
+# `make_dataclass` in the future)
+assert_type(D.__mro__, tuple[type, ...])
```

Comment:
I feel this is a more direct way to do this:
```suggestion
_: type = D
```